app/services/google_translation_service.rb

```diff
@@ -7,6 +7,38 @@ class TranslationError < StandardError; end
   def initialize(project_id:, key:)
     Rails.logger.info('Connecting to Google Cloud Translate...')
     @translate = Google::Cloud::Translate::V2.new("project_id": project_id, "key": key)
+    @cache = Rails.cache
+  end
+
+  def transform_translations_for_caching(listing_id, keys, translations)
+    listing = Force::ListingService.cached_listing(listing_id)
+    # keys can come from updated_values.keys in the event_subscriber_translate_service
+    # they will be in the same order as the translations because the translation service uses the values from that object
+    # this would make the assumption that every value comes back with 1 translation for each translation language
+    # (we would have a problem if it came back with just 2 in the translation array but were expecting 3 fields)
+    return_value = {}
+    translations.each do |target|
+      target[:translation].each_with_index do |value, i|
+        field = keys[i].to_sym
+        return_value[field] ||= {}
+        return_value[field][target[:to].to_sym] = value
+      end
+    end
+    # alternatively, we could splat translations into the listing object
+    # so they aren't nested under the translations key, but instead under each key: {**listing, **return_value}
+    # gets complicated with nested translations (like listing image descriptions)
+    listing[0][:translations] = return_value
```

Comment:
We may also run into the issue of overwriting previous set translations. For example if someone updates "Field A", then we will hit this service to translate that field. Then if someone updates "Field B" we hit this service again but set the translations object to just the result of "Field B".

Using this pattern, we could just do like an update (not exactly this, but similar concept): `listing[0][:translations] = [...translations, return_value[field_to_update]`

And we could also consider batching field updates by setting up a queue.

Reply:
I fixed the overwriting issue you mentioned and added a test case that covers it